service/src/main/java/org/cbioportal/service/impl/AlterationCountServiceImpl.java

```diff
@@ -215,21 +197,87 @@
+                .collect(Collectors.toSet());
 
-        return new Pair<>(alterationCountByGenes, profiledCasesCount);
+            Set<String> molecularProfileIds = updatedProfileCaseIdentifiers
+                .stream()
+                .map(MolecularProfileCaseIdentifier::getMolecularProfileId)
+                .collect(Collectors.toSet());
+            Map<String, String> molecularProfileIdStudyIdMap = molecularProfileRepository
+                .getMolecularProfiles(molecularProfileIds, "SUMMARY")
+                .stream()
+                .collect(Collectors.toMap(MolecularProfile::getStableId, MolecularProfile::getCancerStudyIdentifier));
+
+
+            Map<String, AlterationCountByGene> totalResult = new HashMap<>();
+
+            updatedProfileCaseIdentifiers
+                .stream()
+                .collect(Collectors
+                    .groupingBy(identifier -> molecularProfileIdStudyIdMap.get(identifier.getMolecularProfileId())))
+                .values()
+                .forEach(studyMolecularProfileCaseIdentifiers -> {
+                    List<S> studyAlterationCountByGenes = dataFetcher.apply(studyMolecularProfileCaseIdentifiers);
+                    if (includeFrequency) {
+                        Long studyProfiledCasesCount = includeFrequencyFunction.apply(studyMolecularProfileCaseIdentifiers, studyAlterationCountByGenes);
+                        profiledCasesCount.updateAndGet(v -> v + studyProfiledCasesCount);
+                    }
+                    studyAlterationCountByGenes.forEach(datum -> {
+                        String key = keyGenerator.apply(datum);
+                        if (totalResult.containsKey(key)) {
+                            AlterationCountByGene alterationCountByGene = totalResult.get(key);
+                            alterationCountByGene.setTotalCount(alterationCountByGene.getTotalCount() + datum.getTotalCount());
+                            alterationCountByGene.setNumberOfAlteredCases(alterationCountByGene.getNumberOfAlteredCases() + datum.getNumberOfAlteredCases());
+                            alterationCountByGene.setNumberOfProfiledCases(alterationCountByGene.getNumberOfProfiledCases() + datum.getNumberOfProfiledCases());
+                            Set<String> matchingGenePanelIds = new HashSet<>();
+                            if (CollectionUtils.isNotEmpty(alterationCountByGene.getMatchingGenePanelIds())) {
+                                matchingGenePanelIds.addAll(alterationCountByGene.getMatchingGenePanelIds());
+                            }
+                            if (CollectionUtils.isNotEmpty(datum.getMatchingGenePanelIds())) {
+                                matchingGenePanelIds.addAll(datum.getMatchingGenePanelIds());
+                            }
```

Comment:
These are both guaranteed to be non null
```suggestion
                            if (!alterationCountByGene.getMatchingGenePanelIds().isEmpty()) {
                                matchingGenePanelIds.addAll(alterationCountByGene.getMatchingGenePanelIds());
                            }
                            if (!datum.getMatchingGenePanelIds().isEmpty()) {
                                matchingGenePanelIds.addAll(datum.getMatchingGenePanelIds());
                            }
```

Reply:
They can be null when `includeFrequency` is `false`